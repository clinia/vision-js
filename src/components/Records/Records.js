/** @jsx h */

import { h } from 'preact';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Template from '../Template/Template';

const Records = ({ results, records, cssClasses, templateProps }) => {
  if (results.records.length === 0) {
    return (
      <Template
        {...templateProps}
        templateKey="empty"
        rootProps={{
          className: cx(cssClasses.root, cssClasses.emptyRoot),
        }}
        data={results}
      />
    );
  }

  return (
    <div className={cssClasses.root}>
      <ol className={cssClasses.list}>
        {records.map((record, position) => (
          <Template
            {...templateProps}
            templateKey="item"
            rootTagName="li"
            rootProps={{ className: cssClasses.item }}
            key={record.id}
            data={{
              ...record,
              __recordIndex: position,
            }}
          />
        ))}
      </ol>
    </div>
  );
};

Records.propTypes = {
  cssClasses: PropTypes.shape({
    root: PropTypes.string.isRequired,
    emptyRoot: PropTypes.string.isRequired,
    item: PropTypes.string.isRequired,
    list: PropTypes.string.isRequired,
  }).isRequired,
  records: PropTypes.array.isRequired,
  results: PropTypes.object.isRequired,
  templateProps: PropTypes.object.isRequired,
};

Records.defaultProps = {
  results: { records: [] },
  records: [],
};

export default Records;
