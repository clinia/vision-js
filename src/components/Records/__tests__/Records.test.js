/** @jsx h */

import { h } from 'preact';
import { shallow, mount } from 'enzyme';
import Template from '../../Template/Template';
import Records from '../Records';

describe('Records', () => {
  const cssClasses = {
    root: 'root',
    emptyRoot: 'emptyRoot',
    item: 'item',
    list: 'list',
  };

  function shallowRender(extraProps = {}) {
    const props = {
      cssClasses,
      templateProps: {},
      ...extraProps,
    };

    return shallow(<Records {...props} />);
  }

  describe('no results', () => {
    it('should use the empty template if no results', () => {
      const props = {
        results: {
          records: [],
        },
        records: [],
        cssClasses,
      };

      const wrapper = shallowRender(props);

      expect(wrapper.props().templateKey).toEqual('empty');
    });

    it('should set the empty CSS class when no results', () => {
      const props = {
        results: {
          records: [],
        },
        cssClasses,
      };

      const wrapper = shallowRender(props);

      expect(wrapper.props().rootProps.className).toContain('root');
    });
  });

  describe('individual item templates', () => {
    it('should add an item template for each hit', () => {
      const records = [
        {
          id: 'one',
          foo: 'bar',
        },
        {
          id: 'two',
          foo: 'baz',
        },
      ];
      const props = {
        results: { records },
        records,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find(Template);

      expect(wrapper).toHaveLength(2);
      expect(wrapper.at(0).props().templateKey).toEqual('item');
    });

    it('should set the item class to each item', () => {
      const records = [
        {
          id: 'one',
          foo: 'bar',
        },
      ];
      const props = {
        results: { records },
        records,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find(Template);

      expect(wrapper.props().rootProps.className).toContain('item');
    });

    it('should wrap the items in a root div element', () => {
      const props = {
        results: {
          records: [
            {
              id: 'one',
              foo: 'bar',
            },
            {
              id: 'two',
              foo: 'baz',
            },
          ],
        },
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props);

      expect(wrapper.name()).toEqual('div');
      expect(wrapper.props().className).toContain('root');
    });

    it('should pass each result data to each item template', () => {
      const records = [
        {
          id: 'one',
          foo: 'bar',
        },
        {
          id: 'two',
          foo: 'baz',
        },
      ];
      const props = {
        results: { records },
        records,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find({ templateKey: 'item' });

      expect(wrapper.at(0).props().data.foo).toEqual('bar');
      expect(wrapper.at(1).props().data.foo).toEqual('baz');
    });

    it('should add the __hitIndex in the list to each item', () => {
      const records = [
        {
          id: 'one',
          foo: 'bar',
        },
        {
          id: 'two',
          foo: 'baz',
        },
      ];
      const props = {
        results: { records },
        records,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find({ templateKey: 'item' });

      expect(wrapper.at(0).props().data.__hitIndex).toEqual(0);
      expect(wrapper.at(1).props().data.__hitIndex).toEqual(1);
    });

    it('should use the id as the DOM key', () => {
      const records = [
        {
          id: 'BAR',
          foo: 'bar',
        },
        {
          id: 'BAZ',
          foo: 'baz',
        },
      ];
      const props = {
        results: { records },
        records,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find({ templateKey: 'item' });

      expect(wrapper.at(0).key()).toEqual('BAR');
      expect(wrapper.at(1).key()).toEqual('BAZ');
    });
  });

  describe('markup', () => {
    it('should render <Records />', () => {
      const records = [
        {
          id: 'one',
          foo: 'bar',
        },
        {
          id: 'two',
          foo: 'baz',
        },
      ];

      const props = {
        results: { records },
        records,
        templateProps: {
          templates: {
            item: 'item',
          },
        },
        cssClasses,
      };

      const wrapper = mount(<Records {...props} />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
