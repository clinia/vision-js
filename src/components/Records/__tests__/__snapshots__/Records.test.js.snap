// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Records markup should render <Records /> 1`] = `
Array [
  <div
    className="root"
  >
    <ol
      className="list"
    >
      Array [
        <li
          className="item"
          dangerouslySetInnerHTML={
            Object {
              "__html": "item",
            }
          }
        />,
      ]
      Array [
        <li
          className="item"
          dangerouslySetInnerHTML={
            Object {
              "__html": "item",
            }
          }
        />,
      ]
    </ol>
  </div>,
]
`;

exports[`Records markup should render <Records /> without highlight function 1`] = `
Array [
  <div
    className="root"
  >
    <ol
      className="list"
    >
      Array [
        <li
          className="item"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<mark class=\\"cvi-Highlight-highlighted\\">name 1</mark>",
            }
          }
        />,
      ]
      Array [
        <li
          className="item"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<mark class=\\"cvi-Highlight-highlighted\\">name 2</mark>",
            }
          }
        />,
      ]
    </ol>
  </div>,
]
`;
